arangod/Aql/IResearchViewExecutor.h

```diff
@@ -390,14 +427,44 @@ class IndexReadBuffer {
     StorageSnapshot const* snapshot;
     ValueType value;
   };
+
+  template<typename ColumnReaderProvider>
+  velocypack::Slice readHeapSortColumn(iresearch::HeapSortElement const& cmp,
+                                       irs::doc_id_t doc,
+                                       ColumnReaderProvider& readerProvider,
+                                       size_t readerSlot);
+  template<bool fullHeap, typename ColumnReaderProvider>
+  void finalizeHeapSortDocument(size_t idx, irs::doc_id_t,
+                                std::span<float_t const> scores,
+                                ColumnReaderProvider& readerProvider);
+
   std::vector<BufferValueType> _keyBuffer;
   // FIXME(gnusi): compile time
   std::vector<iresearch::SearchDoc> _searchDocs;
   std::vector<irs::score_t> _scoreBuffer;
-  StoredValuesContainer _storedValuesBuffer;
+  StoredValuesContainer _storedValuesBuffer;  // FIXME: ensure atMost rows are
+                                              // allocated from the beginning!
+
+  // Heap Sort facilities
+  // maps stored column to index in _storedValuesBuffer
+  containers::FlatHashMap<ptrdiff_t, size_t> _heapUsedStoredColumns;
+  // atMost * <num heap only values>
+  StoredValuesContainer _heapOnlyStoredValuesBuffer;
+  // <num heap only values>
+  std::vector<ColumnIterator> _heapOnlyStoredValuesReaders;
+  // <num heap only values>
+  StoredValuesContainer _currentDocumentBuffer;
+  using DocumentSlicesContainer =
+      typename std::conditional<copyStored, bool,
```

Comment:
Why bool? You could write
```suggestion
      IRS_NO_UNIQUE_ADDRESS irs::utils::Need<copyStored, ...>
```

Reply:
Good point